src/stores/index.js

```diff
@@ -851,23 +896,38 @@ const Store = Reflux.createStore({
     // Set the connection's app name to the electron app name of Compass.
     connection.appname = electron.remote.app.getName();
 
-    const dataService = new DataService(connection);
+    if (!this.state.isConnecting) {
+      // The connection attempt might have been cancelled while
+      // we were parsing the connection information, so we return here.
+      return;
+    }
 
     try {
-      const runConnect = promisify(dataService.connect.bind(dataService));
-      const connectedDataService = await runConnect();
+      const dataService = new DataService(connection);
+      this.connectingConnectionAttempt = createConnectionAttempt(dataService);
+
+      const connectedDataService = await this.connectingConnectionAttempt.connect();
+
+      this.connectingConnectionAttempt = null;
+
+      if (!connectedDataService || !this.state.isConnecting) {
+        return;
+      }
 
       const currentConnection = this.state.currentConnection;
```

Comment:
this part and the part below `currentConnection.lastUsed  ...` (which seems to be related),  id really move it in a separate method: `saveEstabilishedConnection` or something similar, is not really clear why this thing is here.

Reply:
Added an `_onConnectSuccess` method. Could break it down a bit more, but I think it's a start.